pkgs/development/tools/godot/default.nix

```diff
@@ -1,103 +1,146 @@
+, libXext
+, libXfixes
+, libXi
 , libXinerama
 , libXrandr
 , libXrender
-, libpulseaudio
-, libXi
-, libXext
-, libXfixes
-, freetype
 , openssl
-, alsa-lib
-, libGLU
-, zlib
+, pkg-config
+, scons
+, stdenv
+, udev
 , yasm
-, withUdev ? true
+, zlib
 }:
 
-let
-  options = {
-    touch = libXi != null;
-    pulseaudio = false;
-    udev = withUdev;
-  };
-in
-stdenv.mkDerivation rec {
+stdenv.mkDerivation (self: {
   pname = "godot";
   version = "3.4.4";
 
   src = fetchFromGitHub {
     owner = "godotengine";
     repo = "godot";
-    rev = "${version}-stable";
-    sha256 = "sha256-3AESLzqozi7Fc80u8Ml3ergZMkIhHy4tNlRe/3FsE6k=";
+    rev = self.passthru.githubRev;
+    sha256 = self.passthru.githubSha256;
+  };
+
+  passthru = {
+    githubRev = "${self.version}-stable";
+    githubSha256 = "sha256-3AESLzqozi7Fc80u8Ml3ergZMkIhHy4tNlRe/3FsE6k=";
+
+    sconsFlagSet = {
+      target = "release_debug";
+      platform = "x11";
+      tools = true;
+      touch = libXi != null;
+      pulseaudio = false;
+      udev = true;
+    };
+
+    buildDescription = null;
   };
 
-  nativeBuildInputs = [ pkg-config ];
+  nativeBuildInputs = [ installShellFiles pkg-config scons ];
+
   buildInputs = [
-    scons
-    udev
+    alsa-lib
+    freetype
+    libGLU
+    libpulseaudio
     libX11
     libXcursor
+    libXext
+    libXfixes
+    libXi
     libXinerama
     libXrandr
     libXrender
-    libXi
-    libXext
-    libXfixes
-    freetype
     openssl
-    alsa-lib
-    libpulseaudio
-    libGLU
-    zlib
+    udev
     yasm
+    zlib
   ];
 
   patches = [ ./pkg_config_additions.patch ./dont_clobber_environment.patch ];
 
   enableParallelBuilding = true;
 
-  sconsFlags = "target=release_debug platform=x11";
-  preConfigure = ''
-    sconsFlags+=" ${
-      lib.concatStringsSep " "
-      (lib.mapAttrsToList (k: v: "${k}=${builtins.toJSON v}") options)
-    }"
-  '';
+  sconsFlags =
+    let
+      toSconsString = x: if builtins.isBool x then lib.boolToString x else toString x;
+    in
+      lib.mapAttrsToList (k: v: "${k}=${toSconsString v}") self.passthru.sconsFlagSet;
 
-  outputs = [ "out" "dev" "man" ];
+  shouldInstallBin = true;
+  shouldInstallShortcut = true;
+  shouldInstallHeaders = true;
+  shouldInstallManual = true;
+
+  outputs = []
+    ++ lib.optional (self.shouldInstallBin || self.shouldInstallShortcut) "out"
+    ++ lib.optional self.shouldInstallHeaders "dev"
+    ++ lib.optional self.shouldInstallManual "man";
+
+  # Where to install the godot bin files within the out directory.
+  binInstallPath = "bin";
+
+  # Pattern identifying the godot executable resulting from the scons build.
+  builtGodotBinNamePattern = "godot.*";
+
+  # Name to give the godot executable when installing in the out path.
+  installedGodotBinName = self.pname;
 
   installPhase = ''
-    mkdir -p "$out/bin"
-    cp bin/godot.* $out/bin/godot
+    runHook preInstall
+
+    if [ $shouldInstallBin ]; then
+      echo "Installing godot binaries."
```

Comment:
Aliasing bin is not great because it is usually used by the bin output. I would recommend to just write it out without a variable.

Reply:
So instead of this:
```sh
bin="$out/$binInstallPath"
mkdir -p "$bin"
cp -R bin/. "$bin"
mv "$bin"/$builtGodotBinNamePattern "$bin/$installedGodotBinName"
```
You'd rather I do this?
```sh
mkdir -p "$out/$binInstallPath"
cp -R bin/. "$out/$binInstallPath"
mv "$out/$binInstallPath"/$builtGodotBinNamePattern "$out/$binInstallPath/$installedGodotBinName"
```
I'm not clear on why ... Seems like unnecessary repetition...

Reply:
We don't need to illuminate every repetition. There needs to be a good middle ground and especially we should not overwrite special nixos variables like $out or the other ouputs.

Reply:
I've kep the variable for this, but renamed it from $bin to $outbin in case something else is using $bin.